store/postgres/src/entities.rs

```diff
@@ -0,0 +1,993 @@
+                let query = format!("
+                    create index if not exists {name} on {subgraph}.entities 
+                    using {index_type} ((data->'{attribute_name}'{jsonb_operator}'data') {index_operator})
+                    where entity='{entity_name}'",
+                    name=name,
+                    subgraph=entities.schema,
+                    index_type=index_type,
+                    attribute_name=&index.attribute_name,
+                    jsonb_operator=jsonb_operator,
+                    index_operator=index_operator,
+                    entity_name=&index.entity_name);
+                conn.batch_execute(&*query)?;
+                Ok(1)
+            }
+        }
+    }
+}
+
+/// Delete all entities. This function exists solely for integration tests
+/// and should never be called from any other code. Unfortunately, Rust makes
+/// it very hard to export items just for testing
+pub fn delete_all_entities(conn: &PgConnection) -> Result<usize, StoreError> {
+    // Delete public entities and related data
+    let mut rows = diesel::delete(public::entities::table).execute(conn)?;
+    rows = rows + diesel::delete(public::entity_history::table).execute(conn)?;
+    rows = rows + diesel::delete(public::event_meta_data::table).execute(conn)?;
+    // Delete subgraphs entities
+    rows = rows + diesel::delete(subgraphs::entities::table).execute(conn)?;
+    Ok(rows)
+}
+
+pub(crate) fn create_schema(
+    conn: &PgConnection,
+    subgraph_id: &SubgraphDeploymentId,
+) -> Result<(), StoreError> {
+    // Create a schema for the deployment
+    let schemas: Vec<String> = diesel::insert_into(sgd_schemas::table)
+        .values(sgd_schemas::subgraph.eq(subgraph_id.to_string()))
+        .returning(sgd_schemas::name)
+        .get_results(conn)?;
+    let schema_name = schemas
+        .first()
+        .ok_or_else(|| format_err!("failed to read schema name for {} back", subgraph_id))?;
+
+    // Note that we have to use conn.batch_execute to issue DDL commands; using
+    // diesel::sql_query(..).execute(conn) can lead to cases where the command
+    // is sent to the database, and shows up in the database logs, but has no
+    // effect at all, but also does not result in an error.
+    let query = format!("create schema {}", schema_name);
+    conn.batch_execute(&*query)?;
+
+    // The order of columns in the primary key matters a lot, since
+    // we want the pk index to also support queries that do not have an id,
+    // just an entity (like counting the number of entities of a certain type)
+    let query = format!(
+        "create table {}.entities(
+            entity       varchar not null,
+            id           varchar not null,
+            data         jsonb,
+            event_source varchar not null,
+            primary key(entity, id)
+        )",
+        schema_name
+    );
+    conn.batch_execute(&*query)?;
+
+    // TODO: should we check that event_source is not "none"?
```

Comment:
Is this an open question still?

Reply:
It is, but I am removing the comment; as the discussion around reverting dynamic data sources shows, there will be a little more work involved with determining the right condition for collecting history, at least for the subgraph of subgraphs. For the other ones, it is fine to always collect history (which is what this comment applied to).